### PR TITLE
PoCL-R: SPIR-V support and other fixes for running buffer-based SYCL apps on top of DPC++

### PR DESCRIPTION
**Pull request by pjaaskel**

### CODE REVIEW

**Comment by eero-t** [comment]
"PoCL-R: Add SPIR-V input support" commit description could also mention the OpenCL 2.0 -> 3.0  change for the minimum/target defines?

**Review by jansol — APPROVED** [review]
Mostly minor or cosmetic stuff with the exception of the assert in shared_cl_context.cc

File: lib/CL/devices/remote/communication.c [hunk 1]
```diff
@@ -2416,7 +2422,7 @@ pocl_network_build_program (remote_device_data_t *ddata, const void *payload,
   POCL_MSG_PRINT_REMOTE ("METADATA SIZE: %zu\n", (size_t)metadata_size);
 
   /* read program binaries */
-  if (!is_binary && !is_builtin)
+  if (!is_binary && !is_builtin && !is_spirv)
```
Comment (franz) [inline lib/CL/devices/remote/communication.c:2425]: Don't we want to read the binaries for SPIR-V compilation ?

**Review by franz — COMMENTED** [review]
Few nitpicks

File: examples/example1/CMakeLists.txt [hunk 2]
```diff
@@ -96,7 +96,8 @@ set_tests_properties( "examples/example1_poclbin"
 
 if(ENABLE_REMOTE_CLIENT AND ENABLE_REMOTE_SERVER AND ENABLE_HOST_CPU_DEVICES)
   add_test(NAME "remote/example1"
-           COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/../../tools/scripts/test_remote_runner_single.sh" "${CMAKE_BINARY_DIR}" "examples/example1/example1")
+           COMMAND "${CMAKE_CURRENT_SOURCE_DIR}/../../tools/scripts/test_remote_runner_single.sh"
```
[the remark this answers is not in the document]
Comment (jansol) [reply examples/example1/CMakeLists.txt:99]: IIRC that breaks if the repo is being used as a git submodule with `add_subdirectory`. Not sure if `ExternalProject_Add` still has the same problem.

File: lib/CL/devices/remote/communication.c (hunk 1 above)
Comment (pjaaskel) [reply lib/CL/devices/remote/communication.c:2425]: Good catch!

**Comment by pjaaskel** [comment]
Thanks. Fixed other remarks except the cmake path suggestion (due to uncertainty from @jansol).

**Comment by jansol** [comment]
Ugh that format-diff.sh change got applied all wrong. Best to not use the suggestion feature I guess.

**Comment by pjaaskel** [comment]
> Ugh that format-diff.sh change got applied all wrong. Best to not use the suggestion feature I guess.

Fixed.

**Comment by jansol** [comment]
The actual script body is still outside of the `pushd`/`popd` pair.

**Comment by jansol** [comment]
The pushd/popd was supposed to be around all the actual commands in case they break if not run in the root directory:

```bash
#!/usr/bin/env bash

GITROOT=$(git rev-parse --show-toplevel 2>/dev/null)
if [ $? -ne 0 ]; then
  echo "must be run in git repo"
  exit 1
fi

# cd to root directory of the git repo
pushd ${GITROOT} > /dev/null

PATCHY=$(mktemp /tmp/pocl.XXXXXXXX.patch)
trap "rm -f $PATCHY" EXIT

git diff -U0 --no-color >$PATCHY

SCRIPTPATH=$( realpath "$0"  )
RELPATH=$(dirname "$SCRIPTPATH")

$RELPATH/clang-format-diff.py -regex '.*(\.h$|\.c$|\.cl$)' -i -p1 -style GNU <$PATCHY
$RELPATH/clang-format-diff.py -regex '(.*(\.hpp$|\.hh$|\.cc$|\.cpp$))|(lib/llvmopencl/.*)|(lib/CL/devices/tce/.*)' -i -p1 -style LLVM <$PATCHY

# cd back wherever we were previously
popd > /dev/null
```

File: examples/example1/CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (jansol) [reply examples/example1/CMakeLists.txt:99]: It seems [PROJECT_SOURCE_DIR](https://cmake.org/cmake/help/latest/variable/PROJECT_SOURCE_DIR.html) could be the thing to use here.

**Review by jansol — APPROVED** [review]
Minor nit: For consistency's sake the format-branch.sh and format-last-commit.sh scripts could get the same `pushd ${GITROOT}`/`popd` wrapping as format-diff.sh (since that makes them work when invoked from anywhere within the repo, rather than requiring to first manually cd to the repo root).

**Comment by pjaaskel** [comment]
> Minor nit: For consistency's sake the format-branch.sh and format-last-commit.sh scripts could get the same `pushd ${GITROOT}`/`popd` wrapping as format-diff.sh (since that makes them work when invoked from anywhere within the repo, rather than requiring to first manually cd to the repo root).

Yep, please add that also to the rest of the scripts in your next PR. It wasn't in the originals and it was "consistently bad" in format-diff.sh before your improvement suggestion :D

**Comment by pjaaskel** [comment]
```
2023-10-06T14:54:02.8263747Z 151/235 Test #1186: conformance_main_integer_ops_micro_3 .........................***Exception: SegFault 98.79 sec
2023-10-06T14:54:02.8266117Z  Initializing random seed to 0.
2023-10-06T14:54:02.8267177Z Requesting Default device based on command line for platform index 0 and device index 0
2023-10-06T14:54:02.8269664Z Compute Device Name = cpu-bdver4-AMD FX-8800P Radeon R7, 12 Compute Cores 4C+8G, Compute Device Vendor = AuthenticAMD, Compute Device Version = OpenCL 3.0 PoCL HSTR: cpu-x86_64-pc-linux-gnu-bdver4, CL C Version = OpenCL C 1.2 PoCL
2023-10-06T14:54:02.8272184Z Device latest conformance version passed: v2022-04-19-01
2023-10-06T14:54:02.8273306Z Supports single precision denormals: YES
2023-10-06T14:54:02.8277395Z sizeof( void*) = 8  (host)
2023-10-06T14:54:02.8278163Z sizeof( void*) = 8  (device)
2023-10-06T14:54:02.8278767Z quick_long_math...
2023-10-06T14:54:02.8280242Z Testing "+" passed
2023-10-06T14:54:02.8281712Z Testing "-" passed
2023-10-06T14:54:02.8282040Z Testing "*" passed
2023-10-06T14:54:02.8282459Z Testing "/" passed
2023-10-06T14:54:02.8282810Z Testing "%" passed
2023-10-06T14:54:02.8283129Z quick_long_math passed
2023-10-06T14:54:02.8283603Z quick_long_logic...
2023-10-06T14:54:02.8283944Z Testing "&" passed
2023-10-06T14:54:02.8284265Z Testing "|" passed
2023-10-06T14:54:02.8284695Z Testing "^" passed
2023-10-06T14:54:02.8285038Z Testing "~" passed
2023-10-06T14:54:02.8285356Z Testing "&&" passed
2023-10-06T14:54:02.8285793Z Testing "||" passed
2023-10-06T14:54:02.8286124Z Testing "!" passed
2023-10-06T14:54:02.8286422Z quick_long_logic passed
2023-10-06T14:54:02.8286876Z quick_ulong_shift...
2023-10-06T14:54:02.8287252Z Testing ">> by vector" passed
2023-10-06T14:54:02.8289199Z Testing "<< by vector" passed
2023-10-06T14:54:02.8289774Z Testing ">> by scalar" passed
2023-10-06T14:54:02.8291505Z Testing "<< by scalar" passed
2023-10-06T14:54:02.8314858Z quick_ulong_shift passed
2023-10-06T14:54:02.8315298Z quick_ulong_compare...
2023-10-06T14:54:02.8315649Z Testing "?:" passed
2023-10-06T14:54:02.8316098Z Testing "<" passed
2023-10-06T14:54:02.8324042Z Testing ">" 
2023-10-06T14:54:02.8324468Z 
```

**Pull request merged**